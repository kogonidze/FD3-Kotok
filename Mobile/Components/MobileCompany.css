.MobileCompanyName
{
  font-weight: bold;
}

.MobileCompanyClients
{
  padding-top: 5px;
}

.Button
{
  margin-right: 5px;
  margin-top: 5px;
}

.Table 
{
  border: double black 3px;
  width: 99%;
  margin-top: 10px;
  border-collapse: collapse;
}

tr.HeadingRow
{
  border: 3px groove black;
}

tr.HeadingRow th
{
  border: 3px double black;
  padding: 5px;
}