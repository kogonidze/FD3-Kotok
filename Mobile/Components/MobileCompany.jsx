import React from 'react';
import PropTypes, { number } from 'prop-types';

import MobileClient from './MobileClient.jsx';
import ClientCard from './ClientCard.jsx';

import './MobileCompany.css';
import { clientsEvents } from './events';

class MobileCompany extends React.PureComponent {

  static propTypes = {
    name: PropTypes.string.isRequired,
    clients:PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        fam: PropTypes.string.isRequired,
        im: PropTypes.string.isRequired,
        otch: PropTypes.string.isRequired,
        balance: PropTypes.number.isRequired,
      })
    ),
  };

  state = {
    name: this.props.name,
    clients: this.props.clients,
    viewMode: 0,            // 0 - просмотр всех, 1 - просмотр активных, 2 - просмотр заблокированных 
    idClientForEdition: 0,            // 0 - режим редактирования отключен, иначе id выбранного для редактирования клиента
  };

  componentDidMount = () => {
    clientsEvents.addListener("EditModeBtnClicked", this.editModeBtnClicked);
    clientsEvents.addListener("CancelEditionBtnClicked", this.cancelEditionBtnClicked);
    clientsEvents.addListener("SaveEditionChangesBtnClicked", this.saveEditionChangesBtnClicked);
  }

  componentWillUnmount = () => {
    clientsEvents.removeAllListeners();
  }

  editModeBtnClicked = (id) => {
    this.setState({idClientForEdition: id});
  }

  cancelEditionBtnClicked = () => {
    this.setState({idClientForEdition: 0});
  }

  saveEditionChangesBtnClicked = (fam, im, otch, balance, id) => {
    var tempArr = this.state.clients.slice();
    var newItem = {fam: fam, im: im, otch: otch, balance: balance, id: id}

    var selectedClientIndex = this.state.clients.findIndex(client => { client.id == id});

    tempArr[selectedClientIndex] = newItem;

    this.setState({clients: tempArr, idClientForEdition: 0});
  }

  setName1 = () => {
    this.setState({name:'МТС'});
  };

  setName2 = () => {
    this.setState({name:'Velcom'});
  };
  
  setViewMode0 = () => {
    this.setState({viewMode: 0});
  }

  setViewMode1 = () => {
    this.setState({viewMode: 1});
  }

  setViewMode2 = () => {
    this.setState({viewMode: 2});
  }

  // setBalance = (clientId,newBalance) => {
  //   let changed=false;
  //   let newClients=[...this.state.clients]; // копия самого массива клиентов
  //   newClients.forEach( (c,i) => {
  //     if ( c.id==clientId && c.balance!=newBalance ) {
  //       let newClient={...c}; // копия хэша изменившегося клиента
  //       newClient.balance=newBalance;
  //       newClients[i]=newClient;
  //       changed=true;
  //     }
  //   } );
  //   if ( changed )
  //     this.setState({clients:newClients});
  // };
  
  render() {

    console.log("MobileCompany render");

    var clientsCode=this.state.clients.map( client => {
      let FIO={fam:client.fam,im:client.im,otch:client.otch};

      if(this.state.viewMode == 0)
        return <MobileClient key={client.id} id={client.id} FIO={FIO} balance={client.balance} />;
      else if(this.state.viewMode == 1 && client.balance >= 0)
        return <MobileClient key={client.id} id={client.id} FIO={FIO} balance={client.balance} />;
      else if(this.state.viewMode == 2 && client.balance < 0)
        return <MobileClient key={client.id} id={client.id} FIO={FIO} balance={client.balance} />;
    });

    var selectedClient;
    if(this.state.idClientForEdition > 0)
    {
      selectedClient = this.state.clients.find(client => {
        if(client.id == this.state.idClientForEdition)
          return client;
      })
    }
    

    return (
      <div className='MobileCompany'>
        <input type="button" value="=МТС" onClick={this.setName1} />
        <input type="button" value="=Velcom" onClick={this.setName2} />
        <div className='MobileCompanyName'>Компания &laquo;{this.state.name}&raquo;</div>
        <div>
          <input type="button" value="Все" onClick={this.setViewMode0} />
          <input type="button" value="Активные" onClick={this.setViewMode1} />
          <input type="button" value="Заблокированные" onClick={this.setViewMode2} />
        </div>
        <div className='MobileCompanyClients'>
          {clientsCode}
        </div>

        {
          (this.state.idClientForEdition != 0) && <ClientCard id={selectedClient.id} fam={selectedClient.fam} im={selectedClient.im} 
          otch={selectedClient.otch} balance={selectedClient.balance} />
        }
        
      </div>
    )
    

  }

}

export default MobileCompany;