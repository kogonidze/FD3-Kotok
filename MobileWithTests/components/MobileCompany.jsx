import React from 'react';
import PropTypes, { number } from 'prop-types';

import MobileClient from './MobileClient.jsx';
import ClientCard from './ClientCard.jsx';

import './MobileCompany.css';
import { clientsEvents } from './events';

class MobileCompany extends React.PureComponent {

  static global_id = 120; 

  static propTypes = {
    name: PropTypes.string.isRequired,
    clients:PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        fam: PropTypes.string.isRequired,
        im: PropTypes.string.isRequired,
        otch: PropTypes.string.isRequired,
        balance: PropTypes.number.isRequired,
      })
    ),
  };

  state = {
    name: this.props.name,
    clients: this.props.clients,
    viewMode: 0,            // 0 - просмотр всех, 1 - просмотр активных, 2 - просмотр заблокированных 
    idClientForEdition: 0,            // 0 - режим редактирования отключен, иначе id выбранного для редактирования клиента
    additionMode: 0,
  };

  componentDidMount = () => {
    clientsEvents.addListener("EditModeBtnClicked", this.editModeOn);
    clientsEvents.addListener("CancelEditionOrAdditionBtnClicked", this.cancelEditionOrAddition);
    clientsEvents.addListener("SaveEditionOrAdditionChangesBtnClicked", this.saveEditionOrAdditionChanges);
    clientsEvents.addListener("DeleteClientBtnClicked", this.deleteClient);
  }

  componentWillUnmount = () => {
    clientsEvents.removeAllListeners();
  }

  editModeOn = (id) => {
    this.setState({idClientForEdition: id});
  }

  cancelEditionOrAddition = () => {
    this.setState({idClientForEdition: 0, additionMode: 0});
  }

  saveEditionOrAdditionChanges = (fam, im, otch, balance, id) => {
    var tempArr = this.state.clients.slice();
    var newItem = {fam: fam, im: im, otch: otch, balance: balance, id: id}

    var selectedClientIndex = this.state.clients.findIndex(client => { if(client.id === id) return client});
    
    if(selectedClientIndex != -1)
    {
      tempArr[selectedClientIndex] = newItem;
    }
    else
    {
      tempArr.push(newItem);
    }
      
    this.setState({clients: tempArr, idClientForEdition: 0, additionMode: 0});
  }

  deleteClient = (id) => {
    // if(confirm("Вы уверены, что хотите удалить клиента?"))
    // {
    //   var tempArr = this.state.clients.slice();

    //   var selectedClientIndex = this.state.clients.findIndex(client => { if(client.id === id) return client});
    //   tempArr.splice(selectedClientIndex,1);
  
    //   this.setState({clients: tempArr});
    // }

    var tempArr = this.state.clients.slice();

    var selectedClientIndex = this.state.clients.findIndex(client => { if(client.id === id) return client});
    tempArr.splice(selectedClientIndex,1);

    this.setState({clients: tempArr});
  }

  setName1 = () => {
    this.setState({name:'МТС'});
  };

  setName2 = () => {
    this.setState({name:'Velcom'});
  };
  
  setViewMode0 = () => {
    this.setState({viewMode: 0});
  }

  setViewMode1 = () => {
    this.setState({viewMode: 1});
  }

  setViewMode2 = () => {
    this.setState({viewMode: 2});
  }

  setAdditionMode = () => {
    this.setState({additionMode: 1});
  }
  
  render() {

    console.log("MobileCompany render");   

    
    var clientsCode=this.state.clients.map( client => {
      if(this.state.viewMode == 0)
        return <MobileClient key={client.id} info={client} />;
      else if(this.state.viewMode == 1 && client.balance >= 0)
        return <MobileClient key={client.id} info={client} />;
      else if(this.state.viewMode == 2 && client.balance < 0)
        return <MobileClient key={client.id} info={client} />;
    });

    var selectedClient;
    if(this.state.idClientForEdition > 0)
    {
      selectedClient = this.state.clients.find(client => {
        if(client.id == this.state.idClientForEdition)
          return client;
      })
    }
    
    return (
      <div className="MobileCompany">
        <input className="Button" type="button" value="=МТС" onClick={this.setName1} />
        <input className="Button" type="button" value="=Velcom" onClick={this.setName2} />
        <div className="MobileCompanyName">Компания &laquo;{this.state.name}&raquo;</div>
        <div>
          <input className="Button" type="button" value="Все" onClick={this.setViewMode0} />
          <input className="Button" type="button" value="Активные" onClick={this.setViewMode1} />
          <input className="Button" type="button" value="Заблокированные" onClick={this.setViewMode2} />
        </div>
        <table className="Table">
          <thead>
            <tr className="HeadingRow">
              <th>Фамилия</th>
              <th>Имя</th>
              <th>Отчество</th>
              <th>Баланс</th>
              <th>Статус</th>
              <th>Редактировать</th>
              <th>Удалить</th>
            </tr>
          </thead>
          <tbody className="Row">
            {clientsCode}
          </tbody>
        </table>
        <div>
          <input className="Button" type="button" value="Добавить клиента" onClick={this.setAdditionMode} />
        </div>
        {
          (this.state.idClientForEdition != 0) && <ClientCard id={selectedClient.id} fam={selectedClient.fam} im={selectedClient.im} 
          otch={selectedClient.otch} balance={selectedClient.balance} />
        }
        {
          (this.state.additionMode != 0) && <ClientCard additionMode={true} id={++MobileCompany.global_id}/>
        }
      </div>
    )
  }
}

export default MobileCompany;