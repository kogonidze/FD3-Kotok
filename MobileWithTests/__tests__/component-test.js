import React from 'react';
import renderer from 'react-test-renderer';
import MobileCompany from '../components/MobileCompany';

let companyName='Velcom';
let clientsArr=[ 
  {id:101, fam:"Иванов", im:"Иван", otch:"Иванович", balance:200}, 
  {id:105, fam:"Сидоров", im:"Сидор", otch:"Сидорович", balance:250}, 
  {id:110, fam:"Петров", im:"Пётр", otch:"Петрович", balance:180},
  {id:120, fam:"Григорьев", im:"Григорий", otch:"Григорьевич", balance:-220},
];

test('работа ExampleComponent', () => {

  const component = renderer.create(
    <MobileCompany name={companyName} clients={clientsArr}/>
  );

  let componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();
  
  const allClientsBtn = component.root.find(el => el.type == "input" && el.props.value == "Все");
  allClientsBtn.props.onClick();
  componentTree = component.toJSON();
  expect(componentTree).toMatchSnapshot();

  const activeClientsBtn = component.root.find(el => el.type == "input" && el.props.value == "Активные");
  activeClientsBtn.props.onClick();
  componentTree = component.toJSON();
  expect(componentTree).toMatchSnapshot();

  const blockedClientsBtn = component.root.find(el => el.type == "input" && el.props.value == "Заблокированные");
  blockedClientsBtn.props.onClick();
  componentTree = component.toJSON();
  expect(componentTree).toMatchSnapshot();

});
