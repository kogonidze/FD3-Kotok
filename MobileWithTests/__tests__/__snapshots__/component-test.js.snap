// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`работа ExampleComponent 1`] = `
<div
  className="MobileCompany"
>
  <input
    className="Button"
    onClick={[Function]}
    type="button"
    value="=МТС"
  />
  <input
    className="Button"
    onClick={[Function]}
    type="button"
    value="=Velcom"
  />
  <div
    className="MobileCompanyName"
  >
    Компания «
    Velcom
    »
  </div>
  <div>
    <input
      className="Button"
      onClick={[Function]}
      type="button"
      value="Все"
    />
    <input
      className="Button"
      onClick={[Function]}
      type="button"
      value="Активные"
    />
    <input
      className="Button"
      onClick={[Function]}
      type="button"
      value="Заблокированные"
    />
  </div>
  <table
    className="Table"
  >
    <thead>
      <tr
        className="HeadingRow"
      >
        <th>
          Фамилия
        </th>
        <th>
          Имя
        </th>
        <th>
          Отчество
        </th>
        <th>
          Баланс
        </th>
        <th>
          Статус
        </th>
        <th>
          Редактировать
        </th>
        <th>
          Удалить
        </th>
      </tr>
    </thead>
    <tbody
      className="Row"
    >
      <tr
        className="ClientInfo"
      >
        <td>
          Иванов
        </td>
        <td>
          Иван
        </td>
        <td>
          Иванович
        </td>
        <td>
          200
        </td>
        <td
          className="StatusActive"
        >
          active
        </td>
        <td>
          <input
            defaultValue="Редактировать"
            onClick={[Function]}
            type="button"
          />
        </td>
        <td>
          <input
            defaultValue="Удалить"
            onClick={[Function]}
            type="button"
          />
        </td>
      </tr>
      <tr
        className="ClientInfo"
      >
        <td>
          Сидоров
        </td>
        <td>
          Сидор
        </td>
        <td>
          Сидорович
        </td>
        <td>
          250
        </td>
        <td
          className="StatusActive"
        >
          active
        </td>
        <td>
          <input
            defaultValue="Редактировать"
            onClick={[Function]}
            type="button"
          />
        </td>
        <td>
          <input
            defaultValue="Удалить"
            onClick={[Function]}
            type="button"
          />
        </td>
      </tr>
      <tr
        className="ClientInfo"
      >
        <td>
          Петров
        </td>
        <td>
          Пётр
        </td>
        <td>
          Петрович
        </td>
        <td>
          180
        </td>
        <td
          className="StatusActive"
        >
          active
        </td>
        <td>
          <input
            defaultValue="Редактировать"
            onClick={[Function]}
            type="button"
          />
        </td>
        <td>
          <input
            defaultValue="Удалить"
            onClick={[Function]}
            type="button"
          />
        </td>
      </tr>
      <tr
        className="ClientInfo"
      >
        <td>
          Григорьев
        </td>
        <td>
          Григорий
        </td>
        <td>
          Григорьевич
        </td>
        <td>
          -220
        </td>
        <td
          className="StatusBlocked"
        >
          blocked
        </td>
        <td>
          <input
            defaultValue="Редактировать"
            onClick={[Function]}
            type="button"
          />
        </td>
        <td>
          <input
            defaultValue="Удалить"
            onClick={[Function]}
            type="button"
          />
        </td>
      </tr>
    </tbody>
  </table>
  <div>
    <input
      className="Button"
      onClick={[Function]}
      type="button"
      value="Добавить клиента"
    />
  </div>
</div>
`;

exports[`работа ExampleComponent 2`] = `
<div
  className="MobileCompany"
>
  <input
    className="Button"
    onClick={[Function]}
    type="button"
    value="=МТС"
  />
  <input
    className="Button"
    onClick={[Function]}
    type="button"
    value="=Velcom"
  />
  <div
    className="MobileCompanyName"
  >
    Компания «
    Velcom
    »
  </div>
  <div>
    <input
      className="Button"
      onClick={[Function]}
      type="button"
      value="Все"
    />
    <input
      className="Button"
      onClick={[Function]}
      type="button"
      value="Активные"
    />
    <input
      className="Button"
      onClick={[Function]}
      type="button"
      value="Заблокированные"
    />
  </div>
  <table
    className="Table"
  >
    <thead>
      <tr
        className="HeadingRow"
      >
        <th>
          Фамилия
        </th>
        <th>
          Имя
        </th>
        <th>
          Отчество
        </th>
        <th>
          Баланс
        </th>
        <th>
          Статус
        </th>
        <th>
          Редактировать
        </th>
        <th>
          Удалить
        </th>
      </tr>
    </thead>
    <tbody
      className="Row"
    >
      <tr
        className="ClientInfo"
      >
        <td>
          Иванов
        </td>
        <td>
          Иван
        </td>
        <td>
          Иванович
        </td>
        <td>
          200
        </td>
        <td
          className="StatusActive"
        >
          active
        </td>
        <td>
          <input
            defaultValue="Редактировать"
            onClick={[Function]}
            type="button"
          />
        </td>
        <td>
          <input
            defaultValue="Удалить"
            onClick={[Function]}
            type="button"
          />
        </td>
      </tr>
      <tr
        className="ClientInfo"
      >
        <td>
          Сидоров
        </td>
        <td>
          Сидор
        </td>
        <td>
          Сидорович
        </td>
        <td>
          250
        </td>
        <td
          className="StatusActive"
        >
          active
        </td>
        <td>
          <input
            defaultValue="Редактировать"
            onClick={[Function]}
            type="button"
          />
        </td>
        <td>
          <input
            defaultValue="Удалить"
            onClick={[Function]}
            type="button"
          />
        </td>
      </tr>
      <tr
        className="ClientInfo"
      >
        <td>
          Петров
        </td>
        <td>
          Пётр
        </td>
        <td>
          Петрович
        </td>
        <td>
          180
        </td>
        <td
          className="StatusActive"
        >
          active
        </td>
        <td>
          <input
            defaultValue="Редактировать"
            onClick={[Function]}
            type="button"
          />
        </td>
        <td>
          <input
            defaultValue="Удалить"
            onClick={[Function]}
            type="button"
          />
        </td>
      </tr>
      <tr
        className="ClientInfo"
      >
        <td>
          Григорьев
        </td>
        <td>
          Григорий
        </td>
        <td>
          Григорьевич
        </td>
        <td>
          -220
        </td>
        <td
          className="StatusBlocked"
        >
          blocked
        </td>
        <td>
          <input
            defaultValue="Редактировать"
            onClick={[Function]}
            type="button"
          />
        </td>
        <td>
          <input
            defaultValue="Удалить"
            onClick={[Function]}
            type="button"
          />
        </td>
      </tr>
    </tbody>
  </table>
  <div>
    <input
      className="Button"
      onClick={[Function]}
      type="button"
      value="Добавить клиента"
    />
  </div>
</div>
`;

exports[`работа ExampleComponent 3`] = `
<div
  className="MobileCompany"
>
  <input
    className="Button"
    onClick={[Function]}
    type="button"
    value="=МТС"
  />
  <input
    className="Button"
    onClick={[Function]}
    type="button"
    value="=Velcom"
  />
  <div
    className="MobileCompanyName"
  >
    Компания «
    Velcom
    »
  </div>
  <div>
    <input
      className="Button"
      onClick={[Function]}
      type="button"
      value="Все"
    />
    <input
      className="Button"
      onClick={[Function]}
      type="button"
      value="Активные"
    />
    <input
      className="Button"
      onClick={[Function]}
      type="button"
      value="Заблокированные"
    />
  </div>
  <table
    className="Table"
  >
    <thead>
      <tr
        className="HeadingRow"
      >
        <th>
          Фамилия
        </th>
        <th>
          Имя
        </th>
        <th>
          Отчество
        </th>
        <th>
          Баланс
        </th>
        <th>
          Статус
        </th>
        <th>
          Редактировать
        </th>
        <th>
          Удалить
        </th>
      </tr>
    </thead>
    <tbody
      className="Row"
    >
      <tr
        className="ClientInfo"
      >
        <td>
          Иванов
        </td>
        <td>
          Иван
        </td>
        <td>
          Иванович
        </td>
        <td>
          200
        </td>
        <td
          className="StatusActive"
        >
          active
        </td>
        <td>
          <input
            defaultValue="Редактировать"
            onClick={[Function]}
            type="button"
          />
        </td>
        <td>
          <input
            defaultValue="Удалить"
            onClick={[Function]}
            type="button"
          />
        </td>
      </tr>
      <tr
        className="ClientInfo"
      >
        <td>
          Сидоров
        </td>
        <td>
          Сидор
        </td>
        <td>
          Сидорович
        </td>
        <td>
          250
        </td>
        <td
          className="StatusActive"
        >
          active
        </td>
        <td>
          <input
            defaultValue="Редактировать"
            onClick={[Function]}
            type="button"
          />
        </td>
        <td>
          <input
            defaultValue="Удалить"
            onClick={[Function]}
            type="button"
          />
        </td>
      </tr>
      <tr
        className="ClientInfo"
      >
        <td>
          Петров
        </td>
        <td>
          Пётр
        </td>
        <td>
          Петрович
        </td>
        <td>
          180
        </td>
        <td
          className="StatusActive"
        >
          active
        </td>
        <td>
          <input
            defaultValue="Редактировать"
            onClick={[Function]}
            type="button"
          />
        </td>
        <td>
          <input
            defaultValue="Удалить"
            onClick={[Function]}
            type="button"
          />
        </td>
      </tr>
    </tbody>
  </table>
  <div>
    <input
      className="Button"
      onClick={[Function]}
      type="button"
      value="Добавить клиента"
    />
  </div>
</div>
`;

exports[`работа ExampleComponent 4`] = `
<div
  className="MobileCompany"
>
  <input
    className="Button"
    onClick={[Function]}
    type="button"
    value="=МТС"
  />
  <input
    className="Button"
    onClick={[Function]}
    type="button"
    value="=Velcom"
  />
  <div
    className="MobileCompanyName"
  >
    Компания «
    Velcom
    »
  </div>
  <div>
    <input
      className="Button"
      onClick={[Function]}
      type="button"
      value="Все"
    />
    <input
      className="Button"
      onClick={[Function]}
      type="button"
      value="Активные"
    />
    <input
      className="Button"
      onClick={[Function]}
      type="button"
      value="Заблокированные"
    />
  </div>
  <table
    className="Table"
  >
    <thead>
      <tr
        className="HeadingRow"
      >
        <th>
          Фамилия
        </th>
        <th>
          Имя
        </th>
        <th>
          Отчество
        </th>
        <th>
          Баланс
        </th>
        <th>
          Статус
        </th>
        <th>
          Редактировать
        </th>
        <th>
          Удалить
        </th>
      </tr>
    </thead>
    <tbody
      className="Row"
    >
      <tr
        className="ClientInfo"
      >
        <td>
          Григорьев
        </td>
        <td>
          Григорий
        </td>
        <td>
          Григорьевич
        </td>
        <td>
          -220
        </td>
        <td
          className="StatusBlocked"
        >
          blocked
        </td>
        <td>
          <input
            defaultValue="Редактировать"
            onClick={[Function]}
            type="button"
          />
        </td>
        <td>
          <input
            defaultValue="Удалить"
            onClick={[Function]}
            type="button"
          />
        </td>
      </tr>
    </tbody>
  </table>
  <div>
    <input
      className="Button"
      onClick={[Function]}
      type="button"
      value="Добавить клиента"
    />
  </div>
</div>
`;
